File: convex/lib/tagCodec.ts
/** OTG scored tags: v1 uint32 IDs, v2 unsigned LEB128 ID deltas.
 * Float64 scores remain lossless. Choose v2 only when it is smaller. */
export function encodeTags(
  entries: Array<[number, number]>,
  version?: 1 | 2,
): ArrayBuffer {
  if (entries.length > 4096) throw new Error("Too many tags");
  const sorted = [...entries].sort((a, b) => a[0] - b[0]);
  let previous = 0;
  const deltas: number[][] = [];
  for (const [code, score] of sorted) {
    if (
      !Number.isInteger(code) ||
      code <= previous ||
      code > 0xffffffff ||
      !Number.isFinite(score) ||
      score < 0 ||
      score > 1
    )
      throw new Error("Invalid tag payload");
    let delta = code - previous;
    const bytes: number[] = [];
    do {
      const byte = delta % 128;
      delta = Math.floor(delta / 128);
      bytes.push(byte + (delta ? 128 : 0));
    } while (delta);
    deltas.push(bytes);
    previous = code;
  }
  const fixedSize = 8 + sorted.length * 12;
  const deltaSize =
    8 + sorted.length * 8 + deltas.reduce((n, d) => n + d.length, 0);
  const selected = version ?? (deltaSize < fixedSize ? 2 : 1);
  const buffer = new ArrayBuffer(selected === 1 ? fixedSize : deltaSize);
  const view = new DataView(buffer);
  view.setUint32(0, 0x4f544700 + selected);
  view.setUint32(4, sorted.length);
  let offset = 8;
  sorted.forEach(([code, score], i) => {
    if (selected === 1) {
      view.setUint32(offset, code);
      offset += 4;
    } else for (const byte of deltas[i]) view.setUint8(offset++, byte);
    view.setFloat64(offset, score);
    offset += 8;
  });
  return buffer;
}

export function decodeTags(buffer: ArrayBuffer): Array<[number, number]> {
  if (buffer.byteLength < 8) throw new Error("Truncated tag payload");
  const view = new DataView(buffer),
    count = view.getUint32(4);
  const version = view.getUint32(0) - 0x4f544700;
  if (
    (version !== 1 && version !== 2) ||
    count > 4096 ||
    (version === 1 && buffer.byteLength !== 8 + count * 12)
  )
    throw new Error("Unknown or invalid tag payload");
  const entries: Array<[number, number]> = [];
  let previous = 0,
    offset = 8;
  for (let i = 0; i < count; i++) {
    let code: number;
    if (version === 1) {
      code = view.getUint32(offset);
      offset += 4;
    } else {
      let delta = 0,
        place = 1;
      for (let n = 0; ; n++) {
        if (n === 5 || offset >= buffer.byteLength)
          throw new Error("Truncated or oversized tag delta");
        const byte = view.getUint8(offset++);
        delta += (byte & 127) * place;
        if (!(byte & 128)) {
          if ((n > 0 && byte === 0) || delta === 0 || delta > 0xffffffff)
            throw new Error("Noncanonical tag delta");
          break;
        }
        place *= 128;
      }
      code = previous + delta;
    }
    if (offset + 8 > buffer.byteLength) throw new Error("Truncated tag score");
    const score = view.getFloat64(offset);
    offset += 8;
    if (
      code <= previous ||
      code > 0xffffffff ||
      !Number.isFinite(score) ||
      score < 0 ||
      score > 1
    )
      throw new Error("Invalid tag payload");
    entries.push([code, score]);
    previous = code;
  }
  if (offset !== buffer.byteLength)
    throw new Error("Trailing tag payload bytes");
  return entries;
}

File: workers/visual/tag_codec.py
"""Lossless scored tags, shared with convex/lib/tagCodec.ts.

OTG v1: uint32 count followed by uint32 code / float64 score pairs, big endian.
OTG v2: same header/count, unsigned LEB128 positive code deltas / float64 scores.
The catalog uses the smaller encoding; readers retain support for both versions.
Codes refer to a shared dictionary, never local list offsets or bit positions.
"""
import math
import struct

MAX_TAGS = 4096


def encode(entries, version=None):
    entries = list(entries)
    if len(entries) > MAX_TAGS:
        raise ValueError('Too many tags')
    if version not in (None, 1, 2):
        raise ValueError('Unknown tag encoding')
    seen = set()
    for code, score in entries:
        if type(code) is not int or not 1 <= code <= 0xffffffff or code in seen:
            raise ValueError('Codes must be unique uint32 identifiers greater than zero')
        if type(score) not in (int, float) or not math.isfinite(score) or not 0 <= score <= 1:
            raise ValueError('Invalid score')
        seen.add(code)
    entries.sort()
    variable = bytearray()
    previous = 0
    for code, score in entries:
        delta = code - previous
        while delta >= 128:
            variable.append((delta & 127) | 128)
            delta >>= 7
        variable.append(delta)
        variable.extend(struct.pack('>d', score))
        previous = code
    version = version or (2 if len(variable) < len(entries) * 12 else 1)
    body = bytes(variable) if version == 2 else b''.join(
        struct.pack('>Id', code, score) for code, score in entries)
    return b'OTG' + bytes([version]) + struct.pack('>I', len(entries)) + body


def decode(payload):
    if len(payload) < 8 or payload[:3] != b'OTG' or payload[3] not in (1, 2):
        raise ValueError('Unknown or truncated tag encoding')
    version = payload[3]
    count, = struct.unpack('>I', payload[4:8])
    if count > MAX_TAGS or (version == 1 and len(payload) != 8 + count * 12):
        raise ValueError('Invalid payload length')
    entries = []
    offset = 8
    previous = 0
    for _ in range(count):
        if version == 1:
            code, = struct.unpack_from('>I', payload, offset)
            offset += 4
        else:
            delta = 0
            for n in range(5):
                if offset >= len(payload):
                    raise ValueError('Truncated tag delta')
                byte = payload[offset]
                offset += 1
                delta += (byte & 127) << (n * 7)
                if not byte & 128:
                    if (n and byte == 0) or not 1 <= delta <= 0xffffffff:
                        raise ValueError('Noncanonical tag delta')
                    break
            else:
                raise ValueError('Oversized tag delta')
            code = previous + delta
        if offset + 8 > len(payload):
            raise ValueError('Truncated tag score')
        score, = struct.unpack_from('>d', payload, offset)
        offset += 8
        if code <= previous:
            raise ValueError('Noncanonical tag order')
        entries.append((code, score))
        previous = code
    if offset != len(payload) or encode(entries, version=version) != payload:
        raise ValueError('Noncanonical payload')
    return entries
